File: server/src/utils/ApiResponse.js
/**
 * @module utils/ApiResponse
 * @description Standardized API response envelope.
 * Every successful response follows the same shape for consistent client parsing.
 *
 * Shape: { status, statusCode, message, data, meta? }
 */
const httpStatus = require('http-status');

class ApiResponse {
  /**
   * @param {number} statusCode
   * @param {string} message
   * @param {*} data
   * @param {Object} [meta] - Pagination, counts, etc.
   */
  constructor(statusCode, message, data = null, meta = null) {
    this.status = 'success';
    this.statusCode = statusCode;
    this.message = message;
    this.data = data;
    if (meta) this.meta = meta;
  }

  /**
   * Sends the response via Express res object.
   * @param {import('express').Response} res
   */
  send(res) {
    return res.status(this.statusCode).json(this);
  }

  // ── Factory Methods ─────────────────────────

  static ok(data, message = 'Success', meta) {
    return new ApiResponse(httpStatus.OK, message, data, meta);
  }

  static created(data, message = 'Resource created') {
    return new ApiResponse(httpStatus.CREATED, message, data);
  }

  static noContent(message = 'No content') {
    return new ApiResponse(httpStatus.NO_CONTENT, message);
  }

  static paginated(data, pagination, message = 'Success') {
    return new ApiResponse(httpStatus.OK, message, data, { pagination });
  }
}

module.exports = ApiResponse;

File: lib/core/network/api_response.dart
import 'package:freezed_annotation/freezed_annotation.dart';

part 'api_response.freezed.dart';
part 'api_response.g.dart';

/// Standardized API response envelope matching the Node.js backend shape.
///
/// Every backend response follows:
/// ```json
/// {
///   "status": "success" | "error",
///   "statusCode": 200,
///   "message": "...",
///   "data": { ... },
///   "meta": { "pagination": { ... } }
/// }
/// ```
@Freezed(genericArgumentFactories: true)
class ApiResponse<T> with _$ApiResponse<T> {
  const factory ApiResponse({
    required String status,
    required int statusCode,
    required String message,
    T? data,
    ApiMeta? meta,
  }) = _ApiResponse;

  factory ApiResponse.fromJson(
    Map<String, dynamic> json,
    T Function(Object?) fromJsonT,
  ) =>
      _$ApiResponseFromJson(json, fromJsonT);
}

@freezed
class ApiMeta with _$ApiMeta {
  const factory ApiMeta({
    PaginationMeta? pagination,
  }) = _ApiMeta;

  factory ApiMeta.fromJson(Map<String, dynamic> json) =>
      _$ApiMetaFromJson(json);
}

@freezed
class PaginationMeta with _$PaginationMeta {
  const factory PaginationMeta({
    required int totalDocs,
    required int totalPages,
    required int page,
    required int limit,
    required bool hasNextPage,
    required bool hasPrevPage,
    int? nextPage,
    int? prevPage,
  }) = _PaginationMeta;

  factory PaginationMeta.fromJson(Map<String, dynamic> json) =>
      _$PaginationMetaFromJson(json);
}

/// API error response shape for structured error handling.
@freezed
class ApiErrorResponse with _$ApiErrorResponse {
  const factory ApiErrorResponse({
    required String status,
    required int statusCode,
    required String message,
    String? code,
    List<FieldError>? errors,
  }) = _ApiErrorResponse;

  factory ApiErrorResponse.fromJson(Map<String, dynamic> json) =>
      _$ApiErrorResponseFromJson(json);
}

@freezed
class FieldError with _$FieldError {
  const factory FieldError({
    required String field,
    required String message,
    String? type,
  }) = _FieldError;

  factory FieldError.fromJson(Map<String, dynamic> json) =>
      _$FieldErrorFromJson(json);
}
